### ISSUE TITLE
W jaki sposób Allegro rozpoznaje produkty jako duplikaty

### ISSUE THREAD

**Issue by pawelosajda** [allegro/allegro-api#8917]
Dzień dobry

Chciałbym uzyskać odpowiedź na temat w jaki sposób Allegro traktuje utworzenie produktu jako duplikat. W dokumentacji pod adresem [dokumentacja](https://developer.allegro.pl/tutorials/jak-jednym-requestem-wystawic-oferte-powiazana-z-produktem-D7Kj9gw4xFA#jak-wystawic-oferte-z-produktem-za-pomoca-zasobu-sale-product-offers) znalazłem poniższą informacje:

![Screenshot from 2024-04-21 11-29-03](https://github.com/allegro/allegro-api/assets/4314822/6626bdd2-71b7-4b09-a0c2-b8490a6117f9)

Dodatkowo chciałbym zaznaczyć, że kod odpowiedzi 409 jest zwracany przez zasób `POST /sale/product-proposals` natomiast zasób `POST /sale/product-offers` nie zwraca takiego kodu odpowiedzi.

Przedstawie w jaki sposób wystawiam oferty.
Do wystawiania ofert korzystam z zasobu `POST /sale/product-offers` nie przekazuje identyfikatora oferty tylko zgodnie z powyższą dokumentacją polegam na przydzieleniu lub utworzeniu produktu przez Allegro czy w takim scenariuszu istnieje możliwość, że przydzielony lub utworzony pośrednio produkt będzie duplikatem?

Poniżej przedstawiam listę trzech produktów które zostały usunięte, czy mógłbym uzyskać informacje jaki był dokładnie powód usunięcia tych produktów i czy to jest przypadek duplikacji? Produkty były przypisane do tej samej oferty.

```
611a115d-1472-4de2-8708-ababdee74a90
db26981a-41dc-4aed-9b0e-3ebbd0d9fe0e
ac2fc71a-034d-4c19-b7c9-fb834e51eab1
```

**Comment by PrzemyslawLukanowski** [comment]
Zestaw reguł, według których identyfikujemy duplikaty, jest zależny od kategorii, w każdej będzie on różny. Zazwyczaj jednak opiera się na kombinacji EAN, wybranych parametrów oraz zdjęć - na tej podstawie sprawdzamy, czy dany produkt istnieje już w naszym katalogu.

> Dodatkowo chciałbym zaznaczyć, że kod odpowiedzi 409 jest zwracany przez zasób POST /sale/product-proposals natomiast zasób POST /sale/product-offers nie zwraca takiego kodu odpowiedzi.

Tak, gdy znajdziemy duplikat produktu w ramach zasobu /sale/product-offer, to automatycznie łączymy go z ofertą.

> Do wystawiania ofert korzystam z zasobu POST /sale/product-offers nie przekazuje identyfikatora oferty tylko zgodnie z powyższą dokumentacją polegam na przydzieleniu lub utworzeniu produktu przez Allegro czy w takim scenariuszu istnieje możliwość, że przydzielony lub utworzony pośrednio produkt będzie duplikatem?

Tak, jest taka możliwość. Produkty weryfikujemy do dwóch godzin od utworzenia, zatem jeśli w polu `product.publication.status` otrzymasz "PROPOSED", oznaczać to będzie, że taki produkt nie został przez nas jeszcze sprawdzony. Więcej na ten temat przeczytasz w [naszym newsie.](https://developer.allegro.pl/news/wyszukiwanie-i-tworzenie-produktow-wprowadzilismy-zmiany-zwiazane-ze-statusem-produktow-g0lbz9V6ds4).

> Poniżej przedstawiam listę trzech produktów które zostały usunięte, czy mógłbym uzyskać informacje jaki był dokładnie powód usunięcia tych produktów i czy to jest przypadek duplikacji? Produkty były przypisane do tej samej oferty.

Z dostępnych mi informacji wynika, że produkty zawierały dwa numery GTIN, który zostały usunięte z naszej strony i to przyczyniło się do skasowania produktów.

**Comment by pawelosajda** [comment]
Mam prośbę o potwierdzenie czy ta oferta jest poprawna czy traktowana jako duplikat? 
```
15537253142
```
Proszę jeszcze jeśli można o sprawdzenie czy na tym koncie występują problemy z duplikatami produktów aktualnie? Jeśli tak to prosiłbym o podesłanie przykładów.

**Comment by PrzemyslawLukanowski** [comment]
Produkt w tej ofercie nie jest duplikatem.

> Proszę jeszcze jeśli można o sprawdzenie czy na tym koncie występują problemy z duplikatami produktów aktualnie? Jeśli tak to prosiłbym o podesłanie przykładów.

Przykro mi, ale nie mam takiej możliwości, aby móc to zweryfikować.

**Comment by pawelosajda** [comment]
Mam jeszcze przykład oferty zakończonej o identyfikatorze `15284679555` która miała przypisany produkt o id `03b438a9-0637-4ed4-bb5c-79525b3e2af8`.
Produkt został zwrócony i wystawiony ponownie z tym samym id produktu `03b438a9-0637-4ed4-bb5c-79525b3e2af8` co powodowało zwrócenie błędu `Wystąpił błąd - Podany produkt tymczasowy jest przypisany do innej oferty.`
Przy po nownym wysłaniu oferty bez identyfikatora produktu został przydzielony produkt o id `b923f299-bf51-4de2-a46c-d6d318da1c22`

produkt `03b438a9-0637-4ed4-bb5c-79525b3e2af8` ma
```
"offerRequirements": {
        "id": "15284679555",
        "parameters": []
    },
```
produkt `b923f299-bf51-4de2-a46c-d6d318da1c22` ma
```
"offerRequirements": {
        "id": "15595309447",
        "parameters": []
    },
```

Czy to jest poprawny sposób korzystania z api? Czy nie jest to traktowane jako duplikat produktu?

**Comment by PrzemyslawLukanowski** [comment]
Tak, jest to prawidłowe zachowanie, ponieważ w tych przypadkach utworzyliśmy produkty tymczasowe, które mogą być połączone wyłącznie z ofertą, na podstawie której zostały utworzone. Takich produktów nie oznaczamy jako duplikaty. Więcej na ich temat przeczytasz w[ tym newsie](https://developer.allegro.pl/news/w-wybranych-kategoriach-bedziemy-tworzyc-produkty-tymczasowe-nn9Y12dbqHM).

**Comment by PrzemyslawLukanowski** [comment]
Witam, produkt cb2d3c48-5253-4834-9060-e3f4f6825e30 to [produkt tymczasowy](https://developer.allegro.pl/news/w-wybranych-kategoriach-bedziemy-tworzyc-produkty-tymczasowe-nn9Y12dbqHM) utworzony na podstawie oferty. Taki produkt może zostać połączony wyłącznie z ofertą, na podstawie której został utworzony. Dlatego udało się utworzyć nowy produkt za pomocą POST /sale/product-proposals.

**Comment by mbcont** [comment]
to co my mamy zrobić? rozwiązanie gdzie produkt jest tymczasowy nie jest idealne, ten produkt też jest  071daee4-51ce-4a03-96b6-82a70914eeff tymczasowy? co potem z duplikatami w ramach jednego konta?

**Comment by PrzemyslawLukanowski** [comment]
> skąd mamy wiedzieć że produkt jest tymczasowy i czy to rodzi jakieś konsekwencje za sobą?

Opisaliśmy to w linkowanym przeze mnie artykule:

**Jak rozpoznasz produkt tymczasowy?**

W szczegółach produktu, które otrzymasz za pomocą [GET /sale/products/{productId}](https://developer.allegro.pl/documentation/#operation/getSaleProduct), dodaliśmy nowe pole “id” w sekcji “offerRequirements”. Zwrócimy w nim identyfikator oferty, na podstawie której go utworzyliśmy - produkt może zostać połączony wyłącznie ze wskazaną ofertą.

Konsekwencją jest - jak wyżej - to, że taki produkt może być połączony z ofertą, na podstawie której został utworzony.

> ten produkt też jest 071daee4-51ce-4a03-96b6-82a70914eeff tymczasowy? 

Nie jest, tak jak wspomniałem ten produkt został utworzony za pomocą POST /sale/product-proposals, a  tymczasem produkty tymczasowe tworzymy po naszej stronie, na podstawie ofert sprzedawców, gdzie m.in. jako parametr "Stan" wybrano wartość inną niż "Nowy".

> co potem z duplikatami w ramach jednego konta?

Produkty tymczasowe nie są traktowane jako duplikaty.

> czy oferta podpięta pod cb2d3c48-5253-4834-9060-e3f4f6825e30 a potem pod 071daee4-51ce-4a03-96b6-82a70914eeff będzie uznawana za duplikat

Taka oferta nie będzie duplikatem, ponieważ wciąż to jest jedna oferta, w której zmianie uległ jedynie przypisany produkt.
